pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,349 @@
+	config, err := parseConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewBucketWithConfig(logger, config, component)
+}
+
+func HTTPMaxConnections(httpMaxConnections oss.HTTPMaxConns) oss.ClientOption {
+	return func(client *oss.Client) {
+		client.Config.HTTPMaxConns = httpMaxConnections
+	}
+}
+
+func NewBucketWithConfig(logger log.Logger, config Config, component string) (*Bucket, error) {
+	if err := validate(config); err != nil {
+		return nil, err
+	}
+
+	userAgent := fmt.Sprintf("thanos-%s", component)
+	httpConnections := oss.HTTPMaxConns{}
+	httpConnections.MaxIdleConns = config.HTTPConfig.MaxIdleConnections
+	httpConnections.MaxIdleConnsPerHost = config.HTTPConfig.MaxIdleConnectionsPerHost
+	var ossLogger log.Logger
+	switch config.LogLevel {
+	case oss.LogOff, oss.Info:
+		ossLogger = level.Info(logger)
+	case oss.Debug:
+		ossLogger = level.Debug(logger)
+	case oss.Error:
+		ossLogger = level.Error(logger)
+	case oss.Warn:
+		ossLogger = level.Warn(logger)
+	}
+
+	client, err := oss.New(
+		config.Endpoint,
+		config.AccessKey,
+		config.SecretKey,
+		oss.UserAgent(userAgent),
+		oss.EnableCRC(config.HTTPConfig.EnableCRC),
+		oss.SetLogLevel(config.LogLevel),
+		oss.Timeout(config.HTTPConfig.ConnectionTimeout, config.HTTPConfig.ReadWriteTimeout),
+		HTTPMaxConnections(httpConnections),
+		oss.SetLogger(slog.New(log.NewStdlibAdapter(ossLogger, log.MessageKey("ossLogger")), "", slog.LstdFlags)))
+	if err != nil {
+		return nil, errors.Wrap(err, "initialize OSS client")
+	}
+
+	bucket := &Bucket{
+		logger: logger,
+		name:   config.Bucket,
+		client: client,
+	}
+	return bucket, nil
+}
+
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+func validate(conf Config) error {
+	if conf.Endpoint == "" {
+		return errors.New("OSS endpoint is empty in config file, please check!")
+	}
+
+	if conf.AccessKey == "" || conf.SecretKey == "" {
+		return errors.New("Access key or secret key is empty in config file, please check!")
+	}
+
+	return nil
+}
+
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	bucket, err := b.client.Bucket(b.name)
+	if err != nil {
+		return errors.Wrap(err, "delete oss object: "+name)
+	}
+
+	return bucket.DeleteObject(name)
+}
+
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	bucket, err := b.client.Bucket(b.name)
+	if err != nil {
+		return false, errors.Wrap(err, "exists oss object: "+name)
+	}
+
+	res, err := bucket.IsObjectExist(name)
+	if err != nil {
+		return false, errors.Wrap(err, "exists oss object")
+	}
+
+	return res, nil
+}
+
+func (b *Bucket) Close() error {
+	return nil
+}
+
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, DirDelim) + DirDelim
+	}
+
+	bucket, err := b.client.Bucket(b.name)
+	if err != nil {
+		return errors.Wrap(err, "iterate oss objects")
+	}
+
+	var marker = oss.Marker("")
+	for {
```

Comment:
In a scenario, if the `ctx` is an instance of context with timeout, then it is timeout, would the `iterator` be stopped? So should we try to listen the channel of `ctx.Done()` to break the iterator? What do you think?